### ISSUE TITLE
Implement Adverb::Eject in core

### ISSUE THREAD

**Issue by lizmat** [Raku/problem-solving#163]
Basically https://modules.raku.org/dist/Adverb::Eject in functionality, but with further optimizations and further pluggability by adding support for `EJECT-KEY` and `EJECT-POS` methods, which would be able to actually get to the low-level implementation and **actually** prevent values from being generated when they are not needed.

**Comment by jnthn** [comment]
I'm not inclined to increase the surface area of the container API for something that should really fall out of optimization. I took a look at what the optimizer is doing; it's already managing to fully inline the appropriate `postcircumfix:{ }` candidate and the `DELETE-KEY` method that is called by that, meaning that in principle it could rip out the lookup.

But it is today? No, and even putting that aside, the `DELETE-KEY` implementation looks rather wasteful. It appears that in the case the value is present it makes no less than *three* hash lookups: an `existskey`, an `atkey`, and a `deletekey`. But `atkey` returns `null` if the key isn't present, so we could simply do the `atkey`, check if it's `null`, and then `deletekey` if so. That means we're down to 2 resolutions in the case it exists, and 1 in the case it doesn't, down from 3 and 2 respectively. But that makes the `deletekey` conditional on the `atkey`, making it harder on the optimizer. Instead we could just make it a sequence of `my \result = nqp::atkey($!storage, $key); nqp::deletekey($!storage, key); nqp::ifnull(result, Any)` or some such, which we should - post inlining - let us kill off the `atkey` part if not needed.

But even that's wasteful, because deep down, looking up the hash bucket to delete from it means we've already done just about all the work to get the value anyway - if only the VM exposed an instruction for doing the get/delete in one. In fact, the fact this is possible is probably a factor in the current design returning the value anyway: it's almost free if things are implemented well enough. We could in theory add a single instruction for that, or we could just have an `atkey`/`deletekey` sequence be spotted by the VM's optimizer and internally transformed into something that finds the bucket once.

In summary, I'm quite confident we can - with modest effort - make a sunk `:delete` fast anyway, without giving the language user more to worry about.

**Comment by lizmat** [comment]
And what about the `@a[foo]:eject` case?  Are you saying that people should just use `splice` for that?

**Comment by lizmat** [comment]
> we could just have an atkey/deletekey sequence be spotted by the VM's optimizer and internally transformed into something that finds the bucket once.

Which would not cover every other module that does the `Associative` role.  I would therefore like you to re-consider having a `EJECT-POS` and `EJECT-KEY` API addition for `Associative` and `Positional` roles.  Would the VM be able to spot that if the implementation does not use `atkey` and `deletekey` ?

**Comment by lizmat** [comment]
Reduced number of lookups in `Hash.DELETE-KEY` with https://github.com/rakudo/rakudo/commit/9217b1c763

**Comment by vrurg** [comment]
I wonder what was the reason to implement `:delete` on arrays this way in first place? Was it only to make `$v = @a[$i]:delete` possible?

**Comment by lizmat** [comment]
Consistency I think.  Just like `.keys` on an array gives you numbers `0...`.  And to remain consistent with hashes, a deletion should *not* affect the other "keys", so that's why it actually stays in the array.

**Comment by lizmat** [comment]
I thought `:eject` would be much clearer, and relieve the system from having to create the value, even when it is not needed.

**Comment by lizmat** [comment]
One of the reasons I came up with this module is questions like these on SO: https://stackoverflow.com/questions/60404000/delete-an-element-from-a-list-or-array-without-the-any-artefact-in-raku

**Comment by MasterDuke17** [comment]
I think part of people's confusion is thinking that delete means remove
entirely. I suspect that when most people say they want to delete an array
element they mean remove the slot *and* its contents. For example, Python's
`del a[3]` acts like Raku's `@a.splice(3, 1)`, not `@a[3]:delete`. It's
probably way to late to change, but having `@a[3]:delete` actually remove
the slot+contents and
`@a[3]:(reset|nullify|clear|something-like-these-that's-not-delete)`. The
current semantics seem like a Perl holdover that might not be expected
unless you've used Perl before.

**Comment by jnthn** [comment]
Primarily, `:delete` and `:exists` on positionals is there for when dealing with sparse arrays (which could, for example, be implemented using a tree representation).

Further to that, all indexing operations today have a consistent algorithmic order, whichever adverb is provided. I don't think that's worth breaking (at least, not in core).

As noted, we already have a `splice`. I'm not convinced we need another way to do that.
